File: juce_backend/tests/golden/scripts/generate_golden_references.py
#!/usr/bin/env python3
"""
GenerateGoldenReferences.py

Phase 4C: Golden reference generation script
Generates deterministic audio reference files for all instruments

Created: December 30, 2025
Author: Bret Bouchard

Usage:
    python generate_golden_references.py
    python generate_golden_references.py --instrument NexSynth
    python generate_golden_references.py --regenerate
"""

import argparse
import os
import struct
import subprocess
import sys
from pathlib import Path
from datetime import datetime


class WAVWriter:
    """Simple WAV file writer for golden references"""

    @staticmethod
    def write(filename, left_channel, right_channel, sample_rate=48000):
        """
        Write stereo audio data to WAV file

        Args:
            filename: Output WAV filename
            left_channel: Left channel float samples (-1.0 to 1.0)
            right_channel: Right channel float samples (-1.0 to 1.0)
            sample_rate: Sample rate in Hz

        Returns:
            True if successful, False otherwise
        """
        try:
            num_samples = len(left_channel)
            assert len(right_channel) == num_samples, "Channel lengths must match"

            with open(filename, "wb") as f:
                # Write WAV header
                f.write(b"RIFF")
                f.write(struct.pack("<I", 36 + num_samples * 2 * 2))  # File size
                f.write(b"WAVE")
                f.write(b"fmt ")
                f.write(struct.pack("<I", 16))  # fmt chunk size
                f.write(struct.pack("<H", 1))  # PCM audio format
                f.write(struct.pack("<H", 2))  # Stereo
                f.write(struct.pack("<I", sample_rate))
                f.write(struct.pack("<I", sample_rate * 2 * 2))  # Byte rate
                f.write(struct.pack("<H", 4))  # Block align
                f.write(struct.pack("<H", 16))  # Bits per sample
                f.write(b"data")
                f.write(struct.pack("<I", num_samples * 2 * 2))  # Data size

                # Write audio data (16-bit PCM)
                for i in range(num_samples):
                    # Clamp and convert to 16-bit integer
                    left_sample = max(-1.0, min(1.0, left_channel[i]))
                    right_sample = max(-1.0, min(1.0, right_channel[i]))

                    left_int = int(left_sample * 32767.0)
                    right_int = int(right_sample * 32767.0)

                    f.write(struct.pack("<h", left_int))
                    f.write(struct.pack("<h", right_int))

            return True

        except Exception as e:
            print(f"ERROR: Failed to write {filename}: {e}")
            return False


class GoldenReferenceGenerator:
    """Generates golden reference WAV files for all instruments"""

    def __init__(self, output_dir="tests/golden/reference"):
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(parents=True, exist_ok=True)

        # Define golden reference specifications
        self.specs = [
            # NexSynth
            {
                "instrument": "NexSynth",
                "note": 60,
                "velocity": 1.0,
                "blocks": 100,
                "filename": "NexSynth_C4_127.wav",
            },
            {
                "instrument": "NexSynth",
                "note": 60,
                "velocity": 0.5,
                "blocks": 100,
                "filename": "NexSynth_C4_064.wav",
            },
            # SamSampler
            {
                "instrument": "SamSampler",
                "note": 60,
                "velocity": 1.0,
                "blocks": 100,
                "filename": "SamSampler_C4_127.wav",
            },
            {
                "instrument": "SamSampler",
                "note": 60,
                "velocity": 0.5,
                "blocks": 100,
                "filename": "SamSampler_C4_064.wav",
            },
            # LocalGal
            {
                "instrument": "LocalGal",
                "note": 60,
                "velocity": 1.0,
                "blocks": 100,
                "filename": "LocalGal_C4_127.wav",
            },
            {
                "instrument": "LocalGal",
                "note": 60,
                "velocity": 0.5,
                "blocks": 100,
                "filename": "LocalGal_C4_064.wav",
            },
            # KaneMarco
            {
                "instrument": "KaneMarco",
                "note": 60,
                "velocity": 1.0,
                "blocks": 100,
                "filename": "KaneMarco_C4_127.wav",
            },
            {
                "instrument": "KaneMarco",
                "note": 60,
                "velocity": 0.5,
                "blocks": 100,
                "filename": "KaneMarco_C4_064.wav",
            },
            # KaneMarcoAether
            {
                "instrument": "KaneMarcoAether",
                "note": 60,
                "velocity": 1.0,
                "blocks": 100,
                "filename": "KaneMarcoAether_C4_127.wav",
            },
            {
                "instrument": "KaneMarcoAether",
                "note": 60,
                "velocity": 0.5,
                "blocks": 100,
                "filename": "KaneMarcoAether_C4_064.wav",
            },
            # KaneMarcoAetherString
            {
                "instrument": "KaneMarcoAetherString",
                "note": 60,
                "velocity": 1.0,
                "blocks": 100,
                "filename": "KaneMarcoAetherString_C4_127.wav",
            },
            {
                "instrument": "KaneMarcoAetherString",
                "note": 60,
                "velocity": 0.5,
                "blocks": 100,
                "filename": "KaneMarcoAetherString_C4_064.wav",
            },
        ]

    def generate_all(self, instrument_filter=None):
        """Generate all golden reference files"""
        print("\n" + "=" * 50)
        print("  Golden Reference Generator")
        print("  Phase 4C: Deterministic Audio Validation")
        print("=" * 50 + "\n")

        print(f"Generation Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")

        success_count = 0
        total_count = 0

        for spec in self.specs:
            # Filter by instrument if specified
            if instrument_filter and spec["instrument"] != instrument_filter:
                continue

            total_count += 1
            output_path = self.output_dir / spec["filename"]

            if self.generate_reference(spec, output_path):
                success_count += 1

            print()  # Blank line between specs

        # Print summary
        print("=" * 50)
        print("Generation Summary")
        print("=" * 50)
        print(f"Total References: {total_count}")
        print(f"Successfully Generated: {success_count}")
        print(f"Failed: {total_count - success_count}")
        print(f"Success Rate: {100.0 * success_count / total_count:.1f}%")
        print("=" * 50 + "\n")

        if success_count == total_count:
            print("✅ All golden references generated successfully!")
            print("\nNext steps:")
            print("  1. Review generated WAV files in tests/golden/reference/")
            print("  2. Commit reference files to repository")
            print("  3. Run golden tests: ./GoldenTest")
            return 0
        else:
            print("⚠️  Some golden references failed to generate")
            return 1

    def generate_reference(self, spec, output_path):
        """Generate a single golden reference using C++ tool"""
        instrument = spec["instrument"]
        note = spec["note"]
        velocity = spec["velocity"]
        blocks = spec["blocks"]
        filename = spec["filename"]

        print(f"Generating golden reference: {filename}")

        # Check if GenerateGoldenReferences executable exists
        exe_path = Path("GenerateGoldenReferences")
        if not exe_path.exists():
            # Try in build directory
            exe_path = Path("../build_phase4c/GenerateGoldenReferences")

        if not exe_path.exists():
            print(f"  ⚠️  GenerateGoldenReferences executable not found")
            print(f"      Build it first: make GenerateGoldenReferences")
            return False

        # Run C++ tool to generate reference
        try:
            cmd = [str(exe_path)]
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
            print(result.stdout)
            if result.stderr:
                print(result.stderr, file=sys.stderr)
            return result.returncode == 0
        except subprocess.TimeoutExpired:
            print(f"  ❌ Generation timed out")
            return False
        except Exception as e:
            print(f"  ❌ Failed to run generation: {e}")
            return False


def main():
    parser = argparse.ArgumentParser(
        description="Generate golden reference WAV files for deterministic audio testing"
    )
    parser.add_argument(
        "--instrument",
        type=str,
        help="Generate references for specific instrument only",
    )
    parser.add_argument(
        "--output-dir",
        type=str,
        default="tests/golden/reference",
        help="Output directory for reference files",
    )
    parser.add_argument(
        "--regenerate",
        action="store_true",
        help="Regenerate all references (overwrite existing)",
    )

    args = parser.parse_args()

    generator = GoldenReferenceGenerator(output_dir=args.output_dir)
    return generator.generate_all(instrument_filter=args.instrument)


if __name__ == "__main__":
    sys.exit(main())

File: juce_backend/tests/golden/GenerateGoldenReferences.cpp
/*
  ==============================================================================

    GenerateGoldenReferences.cpp
    Created: December 30, 2025
    Author: Bret Bouchard

    Phase 4C: Golden reference generation tool
    Generates deterministic audio reference files for all instruments

  ==============================================================================
*/

#include <memory>
#include <cstring>
#include <cmath>
#include <cstdio>
#include <ctime>

// Pure DSP instrument headers
#include "NexSynthDSP.h"
#include "SamSamplerDSP.h"
#include "LocalGalPureDSP.h"
#include "KaneMarcoPureDSP.h"
#include "KaneMarcoAetherPureDSP.h"
#include "KaneMarcoAetherStringPureDSP.h"

// Simple WAV file writer
struct WAVHeader {
    uint8_t riff[4] = {'R', 'I', 'F', 'F'};
    uint32_t fileSize;
    uint8_t wave[4] = {'W', 'A', 'V', 'E'};
    uint8_t fmt[4] = {'f', 'm', 't', ' '};
    uint32_t fmtSize = 16;
    uint16_t audioFormat = 1;  // PCM
    uint16_t numChannels = 2;
    uint32_t sampleRate = 48000;
    uint32_t byteRate = 48000 * 2 * 2;
    uint16_t blockAlign = 4;
    uint16_t bitsPerSample = 16;
    uint8_t data[4] = {'d', 'a', 't', 'a'};
    uint32_t dataSize;
};

/**
 * @brief WAV file writer
 */
class WAVWriter {
public:
    static bool write(const char* filename, float* leftChannel, float* rightChannel, int numSamples, int sampleRate = 48000) {
        FILE* file = fopen(filename, "wb");
        if (!file) {
            fprintf(stderr, "ERROR: Failed to open %s for writing\n", filename);
            return false;
        }

        // Write WAV header
        WAVHeader header;
        header.sampleRate = sampleRate;
        header.dataSize = numSamples * 2 * 2;  // 2 channels, 16-bit
        header.fileSize = 36 + header.dataSize;

        fwrite(&header, sizeof(WAVHeader), 1, file);

        // Convert and write audio data (16-bit PCM)
        for (int i = 0; i < numSamples; ++i) {
            // Clamp and convert to 16-bit integer
            float leftSample = std::max(-1.0f, std::min(1.0f, leftChannel[i]));
            float rightSample = std::max(-1.0f, std::min(1.0f, rightChannel[i]));

            int16_t leftInt = static_cast<int16_t>(leftSample * 32767.0f);
            int16_t rightInt = static_cast<int16_t>(rightSample * 32767.0f);

            fwrite(&leftInt, sizeof(int16_t), 1, file);
            fwrite(&rightInt, sizeof(int16_t), 1, file);
        }

        fclose(file);
        return true;
    }
};

/**
 * @brief Instrument factory
 */
std::unique_ptr<DSP::InstrumentDSP> createInstrument(const char* name) {
    if (strcmp(name, "NexSynth") == 0) return std::make_unique<DSP::NexSynthDSP>();
    if (strcmp(name, "SamSampler") == 0) return std::make_unique<DSP::SamSamplerDSP>();
    if (strcmp(name, "LocalGal") == 0) return std::make_unique<DSP::LocalGalPureDSP>();
    if (strcmp(name, "KaneMarco") == 0) return std::make_unique<DSP::KaneMarcoPureDSP>();
    if (strcmp(name, "KaneMarcoAether") == 0) return std::make_unique<DSP::KaneMarcoAetherPureDSP>();
    if (strcmp(name, "KaneMarcoAetherString") == 0) return std::make_unique<DSP::KaneMarcoAetherStringPureDSP>();
    return nullptr;
}

/**
 * @brief Generate golden reference for an instrument
 */
bool generateGoldenReference(const char* instrumentName, int midiNote, float velocity, int durationBlocks, const char* outputFilename) {
    printf("Generating golden reference: %s\n", outputFilename);

    auto instrument = createInstrument(instrumentName);
    if (!instrument) {
        fprintf(stderr, "ERROR: Failed to create instrument: %s\n", instrumentName);
        return false;
    }

    // Prepare instrument
    const double sampleRate = 48000.0;
    const int blockSize = 512;
    const int numChannels = 2;
    const int totalSamples = durationBlocks * blockSize;

    instrument->prepare(sampleRate, blockSize);

    // Allocate buffers
    std::vector<float> leftChannel(totalSamples);
    std::vector<float> rightChannel(totalSamples);

    // Trigger note
    instrument->noteOn(midiNote, velocity);

    // Process audio
    float* outputs[2];
    for (int block = 0; block < durationBlocks; ++block) {
        outputs[0] = &leftChannel[block * blockSize];
        outputs[1] = &rightChannel[block * blockSize];

        instrument->process(outputs, numChannels, blockSize);
    }

    // Write to WAV file
    bool success = WAVWriter::write(outputFilename, leftChannel.data(), rightChannel.data(), totalSamples, sampleRate);

    if (success) {
        printf("  ✅ Generated: %s\n", outputFilename);

        // Calculate RMS for verification
        double sumLeft = 0.0, sumRight = 0.0;
        for (int i = 0; i < totalSamples; ++i) {
            sumLeft += leftChannel[i] * leftChannel[i];
            sumRight += rightChannel[i] * rightChannel[i];
        }
        double rmsLeft = std::sqrt(sumLeft / totalSamples);
        double rmsRight = std::sqrt(sumRight / totalSamples);

        printf("  RMS: Left=%.6f, Right=%.6f\n", rmsLeft, rmsRight);
    } else {
        fprintf(stderr, "  ❌ Failed to write: %s\n", outputFilename);
    }

    return success;
}

/**
 * @brief Main entry point
 */
int main(int argc, char* argv[]) {
    printf("\n==============================================\n");
    printf("  Golden Reference Generator\n");
    printf("  Phase 4C: Deterministic Audio Validation\n");
    printf("==============================================\n\n");

    // Print timestamp
    time_t now = time(0);
    printf("Generation Time: %s", ctime(&now));

    // Create output directory
    system("mkdir -p tests/golden/reference");

    int successCount = 0;
    int totalCount = 0;

    // Define golden reference specifications
    struct GoldenSpec {
        const char* instrumentName;
        int midiNote;
        float velocity;
        int durationBlocks;
        const char* filename;
    };

    GoldenSpec specs[] = {
        // NexSynth
        {"NexSynth", 60, 1.0f, 100, "tests/golden/reference/NexSynth_C4_127.wav"},
        {"NexSynth", 60, 0.5f, 100, "tests/golden/reference/NexSynth_C4_064.wav"},

        // SamSampler
        {"SamSampler", 60, 1.0f, 100, "tests/golden/reference/SamSampler_C4_127.wav"},
        {"SamSampler", 60, 0.5f, 100, "tests/golden/reference/SamSampler_C4_064.wav"},

        // LocalGal
        {"LocalGal", 60, 1.0f, 100, "tests/golden/reference/LocalGal_C4_127.wav"},
        {"LocalGal", 60, 0.5f, 100, "tests/golden/reference/LocalGal_C4_064.wav"},

        // KaneMarco
        {"KaneMarco", 60, 1.0f, 100, "tests/golden/reference/KaneMarco_C4_127.wav"},
        {"KaneMarco", 60, 0.5f, 100, "tests/golden/reference/KaneMarco_C4_064.wav"},

        // KaneMarcoAether
        {"KaneMarcoAether", 60, 1.0f, 100, "tests/golden/reference/KaneMarcoAether_C4_127.wav"},
        {"KaneMarcoAether", 60, 0.5f, 100, "tests/golden/reference/KaneMarcoAether_C4_064.wav"},

        // KaneMarcoAetherString
        {"KaneMarcoAetherString", 60, 1.0f, 100, "tests/golden/reference/KaneMarcoAetherString_C4_127.wav"},
        {"KaneMarcoAetherString", 60, 0.5f, 100, "tests/golden/reference/KaneMarcoAetherString_C4_064.wav"},
    };

    const int numSpecs = sizeof(specs) / sizeof(specs[0]);

    // Generate all references
    for (int i = 0; i < numSpecs; ++i) {
        totalCount++;
        if (generateGoldenReference(specs[i].instrumentName, specs[i].midiNote, specs[i].velocity,
                                   specs[i].durationBlocks, specs[i].filename)) {
            successCount++;
        }
        printf("\n");
    }

    // Print summary
    printf("==============================================\n");
    printf("Generation Summary\n");
    printf("==============================================\n");
    printf("Total References: %d\n", totalCount);
    printf("Successfully Generated: %d\n", successCount);
    printf("Failed: %d\n", totalCount - successCount);
    printf("Success Rate: %.1f%%\n", (100.0 * successCount) / totalCount);
    printf("==============================================\n\n");

    if (successCount == totalCount) {
        printf("✅ All golden references generated successfully!\n");
        printf("\nNext steps:\n");
        printf("  1. Review generated WAV files in tests/golden/reference/\n");
        printf("  2. Commit reference files to repository\n");
        printf("  3. Run golden tests: ./GoldenTest\n");
        return 0;
    } else {
        printf("⚠️  Some golden references failed to generate\n");
        return 1;
    }
}
